include/CL/sycl/buffer/detail/buffer.hpp

```diff
@@ -68,15 +69,27 @@ class buffer : public detail::buffer_base,
   */
   boost::multi_array_ref<T, Dimensions> access;
 
-  /// The weak pointer to copy back data on buffer deletion
-  weak_ptr_class<T> final_data;
+  /* How to copy back data on buffer destruction, can be modified with
+     set_final_data( ... )
+   */
+  boost::optional<std::function<void(void)>> final_write_back;
+
+  // Used to store the shared pointer used to create the buffer
+  shared_ptr_class<T> input_shared_pointer;
+
+  // Used to store the unique pointer used to create the buffer
+  std::unique_ptr<T> input_unique_pointer;
+
+  // Used to store the shared pointer used as argument of set_final_data
+  shared_ptr_class<T> final_shared_pointer;
```

Comment:
Do we need this? I have the feeling it has been replaced by ``input_shared_pointer``.